### PR TITLE
[doc] Reorganize documentation on distributed learning (fixes #3596)

### LINKED ISSUE #3596: [RFC] Replace "parallel learning" in docs with "distributed learning"?
LightGBM comes with the ability to use multiple machines for training. This can be done with the CLI, or with integrations like Spark, Kubeflow Fairing, and Dask (#3515 ).

Today, the docs refer to training with multiple machines as "parallel learning".

image

I think that this is not quite precise enough, and can lead to some confusion.

LightGBM has at least two types of parallelism:

* within one process (shared memory), using multithreading with OpenMP
* across multiple processes (possibly on multiple machines, and with distributed data), using either sockets or MPI

https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html#parallel-learning-guide only refers to the second case today.

I think we should rename this guide to "Distributed Learning" and use the word "distributed" everywhere in the documentation that talks about using multiple machines to accomplish model training.

Wanted to open this request for comment before I start making changes. What do you think?

### PR DESCRIPTION
**Pull request by jameslamb**
This PR finishes https://github.com/microsoft/LightGBM/issues/3596 (that work was started in #4000), changing the terminology "parallel learning" or "parallel training" to "distributed learning" and "distributed training".

This also re-organizes the the guide on distributed learning. In this proposed reorganization, the guide would have two top-level sections.

* "How Distributed LightGBM works": information on the distributed training algorithms. This might be expanded in the future with other information specific to distributed training in general (not any particular programming language or tool).
* "Integrations": Information on how to run distributed training.

image

In future PRs (for #3814, for example), I'll propose expanding the "integrations --> Dask" section substantially, with code samples and discussion of Dask-specific topics, like how to choose a good partition size and number of partitions, or how to use the Dask package in an environment constrained by firewall rules (https://github.com/microsoft/LightGBM/pull/3994#issuecomment-782152382).

### CODE REVIEW

File: docs/Parallel-Learning-Guide.rst [hunk 1]
```diff
@@ -42,20 +37,58 @@ These algorithms are suited for different scenarios, which is listed in the foll
 | **#feature is large**   | Feature Parallel  | Voting Parallel |
 +-------------------------+-------------------+-----------------+
 
-More details about these parallel algorithms can be found in `optimization in parallel learning <./Features.rst#optimization-in-parallel-learning>`__.
+More details about these parallel algorithms can be found in `optimization in distributed learning <./Features.rst#optimization-in-distributed-learning>`__.
 
-Build Parallel Version
-----------------------
+Integrations
+------------
```
Comment (jameslamb) [inline docs/Parallel-Learning-Guide.rst:43]: old link

https://lightgbm.readthedocs.io/en/docs-jlamb/Parallel-Learning-Guide.html#build-parallel-version

File: docs/Parallel-Learning-Guide.rst [hunk 2]
```diff
@@ -78,11 +111,13 @@ Then write these IP in one file (assume ``mlist.txt``) like following:
 
 **Note**: For Windows users, need to start "smpd" to start MPI service. More details can be found `here`_.
 
-Run Parallel Learning
----------------------
```
Comment (jameslamb) [inline docs/Parallel-Learning-Guide.rst:82]: old link

https://lightgbm.readthedocs.io/en/docs-jlamb/Parallel-Learning-Guide.html#run-parallel-learning

File: examples/parallel_learning/README.md [hunk 3]
```diff
@@ -1,5 +1,6 @@
-Parallel Learning Example
-=========================
```
Comment (jameslamb) [inline examples/parallel_learning/README.md:2]: old link

https://github.com/microsoft/LightGBM/blob/docs/jlamb/examples/parallel_learning/README.md#parallel-learning-example

File: docs/Parallel-Learning-Guide.rst [hunk 4]
```diff
@@ -1,24 +1,19 @@
-Parallel Learning Guide
-=======================
```
Comment (jameslamb) [inline docs/Parallel-Learning-Guide.rst:2]: old link

https://lightgbm.readthedocs.io/en/docs-jlamb/Parallel-Learning-Guide.html#parallel-learning-guide

**Comment by jameslamb** [comment]
Ok I think this is ready for review! 

Docs builds are passing (https://readthedocs.org/projects/lightgbm/builds/). You can explore this PR's docs at https://lightgbm.readthedocs.io/en/docs-jlamb/.

I believe I've accounted for all old links by adding custom anchors. I left comments on each line of the diff that removes a heading, showing a link using the old anchor. If you notice any others that are missed, please let me know.

File: docs/Parallel-Learning-Guide.rst [hunk 5]
```diff
@@ -1,24 +1,19 @@
...
+------------------------------
 
--  `MMLSpark`_ integrates LightGBM into Apache Spark ecosystem.
-   `The following example`_ demonstrates how easy it's possible to utilize the great power of Spark.
-
--  `Kubeflow Fairing`_ suggests using LightGBM in a Kubernetes cluster.
-   `These examples`_ help to get started with LightGBM in a hybrid cloud environment.
-   Also you can use `Kubeflow XGBoost Operator`_ to train LightGBM model.
-   Please check `this example`_ for how to do this.
+This section describes how distributed learning in LightGBM works. To learn how to do this in various programming languages and frameworks, please see `Integrations <#integrations>`_.
```
Comment (StrikerRUS) [inline docs/Parallel-Learning-Guide.rst:13]: This link doesn't work.

File: docs/Parallel-Learning-Guide.rst [hunk 6]
```diff
@@ -42,20 +37,58 @@ These algorithms are suited for different scenarios, which is listed in the foll
...
-Default build version support parallel learning based on the socket.
+This section describes how to run distributed LightGBM training in various programming languages and frameworks. To learn how distributed learning in LightGBM works generally, please see `How Distributed LightGBM Works <#how-distributed-lightgbm-works>`__.
 
-If you need to build parallel version with MPI support, please refer to `Installation Guide <./Installation-Guide.rst#build-mpi-version>`__.
+Apache Spark
+^^^^^^^^^^^^
+
+Apache Spark users can use `MMLSpark`_ for machine learning workflows with LightGBM. This project is not maintained by LightGBM's maintainers.
+
+See `this MMLSpark example`_ and the `the MMLSpark documentation`_ for additional information on using LightGBM on Spark.
```
Comment (StrikerRUS) [inline docs/Parallel-Learning-Guide.rst:52]:
```suggestion
See `this MMLSpark example`_ and `the MMLSpark documentation`_ for additional information on using LightGBM on Spark.
```

File: docs/Parallel-Learning-Guide.rst [hunk 7]
```diff
@@ -42,20 +37,58 @@ These algorithms are suited for different scenarios, which is listed in the foll
...
+
+.. _Build Parallel Version:
 
 Preparation
------------
+'''''''''''
+
+By default, distributed learning with LightGBM uses socket-based communication.
+
+If you need to build parallel version with MPI support, please refer to `Installation Guide <./Installation-Guide.rst#build-mpi-version>`__.
```
Comment (StrikerRUS) [inline docs/Parallel-Learning-Guide.rst:88]:
```suggestion
If you need to build distributed version with MPI support, please refer to `Installation Guide <./Installation-Guide.rst#build-mpi-version>`__.
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
@jameslamb Nice work! Just few comments below.

File: docs/Parallel-Learning-Guide.rst (hunk 5 above)
Comment (StrikerRUS) [reply docs/Parallel-Learning-Guide.rst:13]: For some reason it is `#id1`!..

Comment (jameslamb) [reply docs/Parallel-Learning-Guide.rst:13]: oh weird, didn't notice that! I'll have to experiment and see if I can fix it

Comment (StrikerRUS) [reply docs/Parallel-Learning-Guide.rst:13]: OK, I see now! Should work now.
```suggestion
This section describes how distributed learning in LightGBM works. To learn how to do this in various programming languages and frameworks, please see `Integrations <#integrations>`__.
```

**Review by StrikerRUS — APPROVED** [review]

**Comment by jameslamb** [comment]
closing and re-opening to retrigger CI, thanks to this infrastructure issue on Azure:

```text
section]This job was abandoned. We have detected that logs from the agent may have not finished uploading. We have included our in-memory record of all log lines uploaded before we lost contact with the agent:
##[section]Starting: Initialize job
Agent name: 'lgbubb5e00003IS'
Agent machine name: 'lgbubb5e00003IS'
Current agent version: '2.182.1'
Agent running as: 'AzDevOps'
Prepare build directory.
Set build variables.
Download all required tasks.
Downloading task: CmdLine (2.182.0)
Downloading task: Bash (3.182.0)
```

Closing and reopening instead of pushing an empty commit because I'm on mobile right now, and there is currently only one other build in progress.

**Review by guolinke — APPROVED** [review]
Thank you

**Pull request merged**